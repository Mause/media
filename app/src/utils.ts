import { String } from 'typescript-string-operations';
import { useState, useEffect } from 'react';
import type { RawAxiosRequestHeaders } from 'axios';
import Axios from 'axios';
import * as RRD from 'react-router-dom';
// import axiosRetry from '@vtex/axios-concurrent-retry';
import type {
  Auth0ContextInterface,
  AuthenticationError,
} from '@auth0/auth0-react';
import { useAuth0 } from '@auth0/auth0-react';
import moment from 'moment';
import * as _ from 'lodash-es';

import { FetchEventTarget } from './fetch_stream';
import type { TV } from './SeasonSelectComponent';
import type { EpisodeResponse } from './streaming';

// axiosRetry(Axios, { retries: 3 });

export type GetResponse<T> = T extends {
  get: { responses: { '200': { content: { 'application/json': unknown } } } };
}
  ? T['get']['responses']['200']['content']['application/json']
  : never;

export function subscribe<T>(
  path: string,
  callback: (a: T) => void,
  error: (e: Error) => void,
  authorization: string,
  end: (() => void) | null = () => null,
): () => void {
  const es = FetchEventTarget(path, {
    headers: new Headers({
      Authorization: `Bearer ${authorization}`,
    }),
  });
  const onerror = (event: Event) => {
    error(event as unknown as Error);
  };
  es.addEventListener('abort', onerror);
  const internal_callback = (event: Event) => {
    callback((event as MessageEvent).data as T);
  };
  es.addEventListener('message', internal_callback);

  return () => {
    es.removeEventListener('close', end);
    es.removeEventListener('abort', onerror);
    es.removeEventListener('message', internal_callback);
  };
}

export function getPrefix() {
  const prefix = import.meta.env.REACT_APP_API_PREFIX;

  if (!prefix) {
    return '';
  } else if (prefix.includes('localhost')) {
    return '';
  } else if (prefix) {
    return `https://${prefix}`;
  }
}

export async function load<T>(
  path: string,
  params?: object,
  headers?: RawAxiosRequestHeaders,
): Promise<T> {
  const url = `${getPrefix()}/api/${path}`;
  console.log('Loading', { url, params });
  const t = await Axios.get<T>(url, {
    params,
    withCredentials: true,
    headers,
  });
  return t?.data;
}

interface Res<T> {
  data?: T;
  done: boolean;
  error?: Error;
}

function isAuthenticationError(e: unknown): e is AuthenticationError {
  return typeof e === 'object' && e != null && 'error' in e;
}

export async function getToken(auth0: Auth0ContextInterface): Promise<string> {
  try {
    return await auth0.getAccessTokenSilently();
  } catch (e) {
    if (isAuthenticationError(e) && e.error === 'missing_refresh_token') {
      await auth0.loginWithRedirect({
        authorizationParams: {
          redirect_uri: window.location.toString(),
        },
      });
      return '';
    } else {
      throw e;
    }
  }
}

export function usePost<T>(
  url: string,
  body: object,
): { done: boolean; data?: T; error?: Error } {
  const [result, setResult] = useState<Res<T>>({ done: false });
  const auth = useAuth0();

  useEffect(() => {
    const abortController = new AbortController();
    getToken(auth)
      .then((token) =>
        Axios.post<T>(`/api/${url}`, body, {
          signal: abortController.signal,
          headers: {
            Authorization: `Bearer ${token}`,
          },
        }),
      )
      .then(
        (res) => {
          setResult({ done: true, data: res.data });
        },
        (error: unknown) => {
          setResult({ done: true, error: error as Error });
        },
      );

    return () => {
      abortController.abort();
    };
  }, [url, body, auth]);

  return result;
}

export function useLocation<T>() {
  const location = RRD.useLocation();
  return { ...location, state: location.state as T };
}

export function getMarker(episode: {
  season?: number;
  episode?: number | null;
}) {
  if (episode.episode) {
    return String.format('S{0:00}E{1:00}', episode.season, episode.episode);
  } else {
    return String.format('S{0:00}', episode.season);
  }
}

export function getMessage(air_date: string) {
  const today = moment().startOf('day');
  const tomorrow = today.clone().add(1, 'day');
  const yesterday = today.clone().subtract(1, 'day');
  const dt = moment(air_date);
  const dts = dt.format('DD/MM/YYYY');

  let message;
  if (today.isSame(dt)) {
    message = 'airs today';
  } else if (dt.isSame(yesterday)) {
    message = 'aired yesterday';
  } else if (dt.isSame(tomorrow)) {
    message = 'airs tomorrow';
  } else if (dt.isAfter(today)) {
    message = `airs on ${dts}`;
  } else {
    message = `aired on ${dts}`;
  }
  return message;
}

export function shouldCollapse(
  i: string,
  data: TV | undefined,
  episodes: EpisodeResponse[],
): boolean {
  let collapse = false;
  if (data) {
    const i_i = +i;
    const seasonMeta = data.seasons.find((s) => s.season_number === i_i);
    if (seasonMeta) {
      const hasNext = true; // !!data.seasons[i_i + 1];

      const episodeNumbers = _.range(1, seasonMeta.episode_count + 1);
      const hasNumbers = _.map(episodes, 'episode');
      const hasAllEpisodes =
        _.difference(episodeNumbers, hasNumbers).length === 0;
      collapse = hasNext && hasAllEpisodes;
    }
  }
  return collapse;
}
