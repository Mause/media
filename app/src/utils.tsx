import React, { ReactElement } from 'react';
import Axios from 'axios';
import { useState, useEffect } from 'react';
import MaterialLink from '@material-ui/core/Link';
import { Link } from 'react-router-dom';
import { LocationDescriptor } from 'history';
import axiosRetry from '@vtex/axios-concurrent-retry';
import { TypographyTypeMap } from '@material-ui/core';
import moxios from 'moxios';
import { unstable_batchedUpdates } from 'react-dom';

// axiosRetry(Axios, { retries: 3 });

export function MLink<S>(props: {
  children: React.ReactNode;
  to: LocationDescriptor<S>;
  color?: TypographyTypeMap['props']['color'];
}): ReactElement {
  return <MaterialLink component={Link} {...props} />;
}

export function subscribe<T>(
  path: string,
  callback: (a: T) => void,
  end: (() => void) | null = null,
): void {
  const es = new EventSource(path, {
    withCredentials: true,
  });
  es.addEventListener('message', ({ data }) => {
    if (!data) {
      if (end) {
        end();
      }
      return es.close();
    }
    callback(JSON.parse(data));
  });
}

export async function load<T>(path: string, params?: string): Promise<T> {
  const t = await Axios.get<T>(`/api/${path}`, {
    params,
    withCredentials: true,
  });
  return t && t.data;
}

export function usePost<T>(
  url: string,
  body: object,
): { done: boolean; data?: T; error?: Error } {
  const [done, setDone] = useState<boolean>(false);
  const [error, setError] = useState<Error>();
  const [data, setData] = useState<T>();

  useEffect(
    () => {
      Axios.post<T>('/api/' + url, body, {
        withCredentials: true,
      }).then(
        res => {
          unstable_batchedUpdates(() => {
            setDone(true);
            setData(res.data);
          });
        },
        error => {
          unstable_batchedUpdates(() => {
            setDone(true);
            setError(error);
          });
        },
      );
    },
    [url, body],
  );

  return { done, error, data };
}

export function ExtMLink(props: { href: string; children: string }) {
  return (
    <MaterialLink href={props.href} target="_blank" rel="noopener noreferrer">
      {props.children}
    </MaterialLink>
  );
}

export function expectLastRequestBody() {
  return expect(JSON.parse(moxios.requests.mostRecent().config.data));
}
