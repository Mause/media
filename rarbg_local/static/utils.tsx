import Axios from 'axios';
import { useState, useEffect } from 'react';

export function subscribe(path: string, callback: (a: any) => void, end: (() => void) | null = null): void {
  const es = new EventSource(path, {
    withCredentials: true,
  });
  es.addEventListener('message', ({ data }) => {
    if (!data) {
      if (end) {
        end();
      }
      return es.close();
    }
    callback(JSON.parse(data));
  });
}

export function load<T>(path: string, params?: any): Promise<T> {
  return Axios.get<T>(`/api/${path}`, { params, withCredentials: true }).then(t => t.data);
}

export function useLoad<T>(path: string, params: any = null) {
  const [data, setData] = useState<T>()
  useEffect(() => {
    load<T>(path, params).then(setData);
  }, [path, params]);
  return data;
}
