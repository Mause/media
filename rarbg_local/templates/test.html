{% extends "base.html" %}

{% block header %}
Options for "{{title}}"
{% endblock %}

{% block content %}
<script type="text/javascript">
  var require = {
    bundles: JSON.parse(`{{BUNDLES|safe}}`),
    deps: ['streaming'],
    callback: function () {
      console.log('Loaded');
    }
  };
</script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.js"></script>

<div id="app"></div>
{% endblock %}
