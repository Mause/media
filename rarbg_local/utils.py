from functools import lru_cache as _lru_cache
from typing import Set

_caches: Set[_lru_cache] = set()  # type: ignore


def lru_cache(*args, **kwargs):
    def wrapper(func):
        cache = _lru_cache(*args, **kwargs)(func)
        _caches.add(cache)
        return cache

    return wrapper


def cache_clear():
    while _caches:
        cache = _caches.pop()
        cache.cache_clear()  # type: ignore
