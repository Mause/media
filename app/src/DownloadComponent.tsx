import { Navigate } from 'react-router-dom';
import type { ReactNode } from 'react';

import { usePost, useLocation } from './utils';
import type { components } from './schema';
import { DisplayError } from './DisplayError';
import { Loading } from './render';
import { RouteTitle } from './RouteTitle';

export type DownloadCall = components['schemas']['DownloadPost'];
export interface DownloadState {
  downloads: DownloadCall[];
}

export function DownloadComponent() {
  const { state } = useLocation<DownloadState>();

  const { done, error } = usePost(
    'download',
    state.downloads.map((item: DownloadCall) => ({
      ...item,
      tmdb_id: Number(item.tmdb_id),
    })),
  );

  console.log({ done, error });

  let res: ReactNode;
  if (done) {
    res = <Navigate to="/" />;
  } else if (error) {
    res = <DisplayError error={error} />;
  } else {
    res = <Loading loading />;
  }

  return <RouteTitle title="Download">{res}</RouteTitle>;
}
