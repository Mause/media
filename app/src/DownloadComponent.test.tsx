import { act } from '@testing-library/react';
import React from 'react';
import { DownloadComponent, DownloadState } from './DownloadComponent';
import { Route, Router } from 'react-router-dom';
import { wait, useMoxios, renderWithSWR } from './test.utils';
import { createMemoryHistory } from 'history';
import moxios from 'moxios';

useMoxios();

test('DownloadComponent', async () => {
  await act(async () => {
    const history = createMemoryHistory();
    const state: DownloadState = {
      downloads: [
        {
          tmdb_id: '10000',
          magnet: '...',
        },
      ],
    };
    history.push('/download', state);

    const el = renderWithSWR(
      <Router history={history}>
        <Route path="/download">
          <DownloadComponent />
        </Route>
      </Router>,
    );

    expect(el.container).toMatchSnapshot();

    await moxios.stubOnce('POST', /\/api\/download/, {});
    await wait();

    expect(el.container).toMatchSnapshot();
    expect(history.length).toBe(2);
  });
});
