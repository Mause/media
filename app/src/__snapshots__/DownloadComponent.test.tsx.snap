// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DownloadComponent success 1`] = `
<div>
  <div
    style="fill: #000000; height: 64px; width: 64px;"
  >
    <svg
      class="icon-loading"
      viewBox="0 0 32 32"
      xmlns="http://www.w3.org/2000/svg"
    >
      
  
      <path
        d="M4 12 A4 4 0 0 0 4 20 A4 4 0 0 0 4 12"
        transform="translate(-8 0)"
      >
         
    
        <animatetransform
          attributeName="transform"
          begin="0"
          calcMode="spline"
          dur="0.8s"
          keySplines="0.2 0.2 0.4 0.8;0.2 0.6 0.4 0.8"
          keyTimes="0;.25;1"
          repeatCount="indefinite"
          type="translate"
          values="-8 0; 2 0; 2 0;"
        />
        
  
      </path>
      
  
      <path
        d="M4 12 A4 4 0 0 0 4 20 A4 4 0 0 0 4 12"
        transform="translate(2 0)"
      >
         
    
        <animatetransform
          attributeName="transform"
          begin="0"
          calcMode="spline"
          dur="0.8s"
          keySplines="0.2 0.2 0.4 0.8;0.2 0.6 0.4 0.8"
          keyTimes="0;.35;1"
          repeatCount="indefinite"
          type="translate"
          values="2 0; 12 0; 12 0;"
        />
        
  
      </path>
      
  
      <path
        d="M4 12 A4 4 0 0 0 4 20 A4 4 0 0 0 4 12"
        transform="translate(12 0)"
      >
         
    
        <animatetransform
          attributeName="transform"
          begin="0"
          calcMode="spline"
          dur="0.8s"
          keySplines="0.2 0.2 0.4 0.8;0.2 0.6 0.4 0.8"
          keyTimes="0;.45;1"
          repeatCount="indefinite"
          type="translate"
          values="12 0; 22 0; 22 0;"
        />
        
  
      </path>
      
  
      <path
        d="M4 12 A4 4 0 0 0 4 20 A4 4 0 0 0 4 12"
        transform="translate(24 0)"
      >
         
    
        <animatetransform
          attributeName="transform"
          begin="0"
          calcMode="spline"
          dur="0.8s"
          keySplines="0.2 0.2 0.4 0.8;0.2 0.6 0.4 0.8"
          keyTimes="0;.55;1"
          repeatCount="indefinite"
          type="translate"
          values="22 0; 32 0; 32 0;"
        />
        
  
      </path>
      

    </svg>
    

  </div>
</div>
`;

exports[`DownloadComponent success 2`] = `<div />`;
